### PR TITLE
Handle event home screen

### PR DESCRIPTION
**Pull request by ndt0409**
## Related Tickets
- [#Task 55573: comic screen](https://edu-redmine.sun-asterisk.vn/issues/55573)
- [#Task 55574: favorite screen](https://edu-redmine.sun-asterisk.vn/issues/55574)
- [#Task 55575: handle event home screen](https://edu-redmine.sun-asterisk.vn/issues/55575)
## WHAT
- create ui comic screen 
- create ui favorite screen
- create request api
- show event 
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/ndt/marvelapp/base/BaseRepository.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import kotlinx.coroutines.withContext
+import java.io.IOException
+
+open class BaseRepository {
+    suspend fun <T> getResult(request: suspend CoroutineScope.() -> T): DataResult<T> {
+        return withContext(Dispatchers.IO) {
+            try {
+                DataResult.Success(request())
+            } catch (e: IOException) {
+                Log.v("error", "Exception=[$e]")
```
Comment (hungld-1302) [inline app/src/main/java/com/ndt/marvelapp/base/BaseRepository.kt]: remove

File: app/src/main/res/layout/fragment_comic.xml [hunk 2]
```diff
@@ -1,8 +1,40 @@
...
 <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
     xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    tools:context=".ui.comic.ComicFragment">
+    android:layout_height="match_parent">
+
+    <Button
+        android:id="@+id/button_search"
```
Comment (hungld-1302) [inline app/src/main/res/layout/fragment_comic.xml]: btn_search

File: app/src/main/res/layout/fragment_list_comic.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".ui.listcomic.ListComicFragment">
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungld-1302) [inline app/src/main/res/layout/fragment_list_comic.xml]: endline

File: app/src/main/res/layout/fragment_see_more.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".ui.seemore.SeeMoreFragment">
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungld-1302) [inline app/src/main/res/layout/fragment_see_more.xml]: .

File: app/src/main/res/values/dimens.xml [hunk 5]
```diff
@@ -5,13 +5,13 @@
...
     <dimen name="dp_24">24dp</dimen>
+    <dimen name="dp_40">40dp</dimen>
     <dimen name="dp_50">50dp</dimen>
     <dimen name="dp_100">100dp</dimen>
     <dimen name="dp_125">125dp</dimen>
 
     <dimen name="sp_14">14sp</dimen>
     <dimen name="sp_16">16sp</dimen>
-    <dimen name="dp_40">40dp</dimen>
     <dimen name="sp_12">12sp</dimen>
```
Comment (hungld-1302) [inline app/src/main/res/values/dimens.xml]: sort

File: app/src/main/java/com/ndt/marvelapp/ui/adapter/OnClickListener.kt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.ndt.marvelapp.ui.adapter
+
+import android.view.View
+
+interface OnClickListener<T> {
+    fun onClick(data: T) {}
+
+    fun onLongClick(data: T) {}
+
+    fun onItemClick(data: T) {}
```
Comment (hungld-1302) [inline app/src/main/java/com/ndt/marvelapp/ui/adapter/OnClickListener.kt]: merge onItemClick(view: View =null, data: T = null)

File: app/src/main/java/com/ndt/marvelapp/ui/home/HomeViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+class HomeViewModel(
+    private val eventRepository: EventRepository,
+    private val creatorRepository: CreatorRepository
+) : BaseViewModel() {
+
+    private val _event = MutableLiveData<List<Event>>()
+    val event: LiveData<List<Event>> get() = _event
+
+    private val _creator = MutableLiveData<List<com.ndt.marvelapp.data.model.Creator>>()
+    val creator: LiveData<List<com.ndt.marvelapp.data.model.Creator>> get() = _creator
```
Comment (hungvv-1301) [inline app/src/main/java/com/ndt/marvelapp/ui/home/HomeViewModel.kt]: sửa lại impl  hoặc dùng type alias a zai

**Review by hungld-1302 — APPROVED** [review]

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -33,4 +47,4 @@
         </activity>
     </application>
 
-</manifest>
+</manifest>
```
Comment (ghost) [inline app/src/main/AndroidManifest.xml]: endline a ơi

File: app/src/main/java/com/ndt/marvelapp/utils/ImageExtension.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import com.bumptech.glide.Glide
+import com.ndt.marvelapp.R
+import com.ndt.marvelapp.utils.BaseUrlImage.urlImage
+
+fun ImageView.loadImage(context: Context, image: String?) {
+    Glide.with(context).load(image?.let { urlImage(it) })
+        .error(R.drawable.ic_error)
+        .placeholder(R.drawable.ic_place_holder)
+        .into(this)
+}
```
Comment (ghost) [inline app/src/main/java/com/ndt/marvelapp/utils/ImageExtension.kt]: dùng context của image view nhá a

**Review by ghost — APPROVED** [review]

File: app/src/main/java/com/ndt/marvelapp/data/model/Creator.kt [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+data class Creator(
+    @SerializedName("id")
+    val id: Int,
+    @SerializedName("fullName")
+    val fullName: String,
+    @SerializedName("modified")
+    val modified: String,
+    @SerializedName("thumbnail")
+    val thumbnail: Image
```
Comment (vietnh-1109) [inline app/src/main/java/com/ndt/marvelapp/data/model/Creator.kt]: nhớ check null ? , gson nó parse null là crash đấy Tú

File: app/src/main/java/com/ndt/marvelapp/data/model/Creator.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+) {
+
+    companion object {
+        val diffUtil = object : DiffUtil.ItemCallback<Creator>() {
+            override fun areItemsTheSame(oldItem: Creator, newItem: Creator) =
+                oldItem.id == newItem.id
+
+            override fun areContentsTheSame(oldItem: Creator, newItem: Creator) = oldItem == newItem
```
Comment (vietnh-1109) [inline app/src/main/java/com/ndt/marvelapp/data/model/Creator.kt]:
```suggestion
            override fun areContentsTheSame(oldItem: Creator, newItem: Creator) = 
                    oldItem == newItem
```

File: app/src/main/java/com/ndt/marvelapp/ui/adapter/EventAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        private var data: Event? = null
+
+        init {
+            binding.root.setOnClickListener(this)
+        }
+
+        override fun onBindData(data: Event) {
+            binding.textEvent.text = data.title
+            binding.imageEvents.loadImage(data.thumbnail)
+            this.data = data
```
Comment (vietnh-1109) [inline app/src/main/java/com/ndt/marvelapp/ui/adapter/EventAdapter.kt:35]: đưa lên đầu đi ông

**Pull request merged**